### PR TITLE
Fix: parallel replicas with PREWHERE (ubsan)

### LINKED ISSUE #60762: UBSAN: InterpreterSelectQuery.cpp:2110:51: runtime error: member call on null pointer of type 'DB::ActionsDAG'
https://s3.amazonaws.com/clickhouse-test-reports/60672/57482de0f0858f5a03e60a7310227106487ce438/ast_fuzzer__ubsan_.html

```
/build/src/Interpreters/InterpreterSelectQuery.cpp:2110:51: runtime error: member call on null pointer of type 'DB::ActionsDAG'
    #0 0x56294252965e in DB::InterpreterSelectQuery::applyFiltersToPrewhereInAnalysis(DB::ExpressionAnalysisResult&) const build_docker/./src/Interpreters/InterpreterSelectQuery.cpp:2110:51
    #1 0x5629425233d3 in DB::InterpreterSelectQuery::addPrewhereAliasActions() build_docker/./src/Interpreters/InterpreterSelectQuery.cpp:2118:5
    #2 0x562942514672 in DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context> const&, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<DB::IS
torage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_
traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::PreparedSets>) build_docker/./src/Interpreters/InterpreterSelectQuery.cpp:
855:9
    #3 0x562942508695 in DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const> const&, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<
DB::IStorage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1:
:char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::PreparedSets>) build_docker/./src/Interpreters/InterpreterSelectQuer
y.cpp:364:7
    #4 0x562942508432 in DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&, std::__1::vector<std
::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./s
rc/Interpreters/InterpreterSelectQuery.cpp:204:7
    #5 0x562941dc6257 in DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher::visit(DB::ASTSelectQuery&, std::__1::shared_ptr<DB::IAST>&, DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher::Data&) build_do
cker/./src/Interpreters/InterpreterExplainQuery.cpp:72:36
    #6 0x562941dc6257 in DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher::visit(std::__1::shared_ptr<DB::IAST>&, DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher::Data&) build_docker/./src/Interprete
rs/InterpreterExplainQuery.cpp:65:17
    #7 0x562941dc6257 in DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::doVisit(std::__1::shared_ptr<DB::IAST>&) build_docker/./src/Int
erpreters/InDepthNodeVisitor.h:71:13
    #8 0x562941dc69d3 in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImplMain<false>(std::__1::shared_ptr<DB::IAST>&) buil
d_docker/./src/Interpreters/InDepthNodeVisitor.h:61:9
    #9 0x562941dc69d3 in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImpl<false>(std::__1::shared_ptr<DB::IAST>&) build_do
cker/./src/Interpreters/InDepthNodeVisitor.h:51:13
    #10 0x562941dc69d3 in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitChildren<false>(std::__1::shared_ptr<DB::IAST>&) bui
ld_docker/./src/Interpreters/InDepthNodeVisitor.h:92:17
    #11 0x562941dc69de in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImplMain<false>(std::__1::shared_ptr<DB::IAST>&) bui
ld_docker/./src/Interpreters/InDepthNodeVisitor.h:64:13
    #12 0x562941dc69de in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImpl<false>(std::__1::shared_ptr<DB::IAST>&) build_d
ocker/./src/Interpreters/InDepthNodeVisitor.h:51:13
    #13 0x562941dc69de in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitChildren<false>(std::__1::shared_ptr<DB::IAST>&) bui
ld_docker/./src/Interpreters/InDepthNodeVisitor.h:92:17
    #14 0x562941dc69de in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImplMain<false>(std::__1::shared_ptr<DB::IAST>&) bui
ld_docker/./src/Interpreters/InDepthNodeVisitor.h:64:13
    #15 0x562941dc69de in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImpl<false>(std::__1::shared_ptr<DB::IAST>&) build_d
ocker/./src/Interpreters/InDepthNodeVisitor.h:51:13
    #16 0x562941dc69de in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitChildren<false>(std::__1::shared_ptr<DB::IAST>&) bui
ld_docker/./src/Interpreters/InDepthNodeVisitor.h:92:17
    #17 0x562941dbca04 in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImplMain<false>(std::__1::shared_ptr<DB::IAST>&) bui
ld_docker/./src/Interpreters/InDepthNodeVisitor.h:64:13
    #18 0x562941dbca04 in void DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visitImpl<false>(std::__1::shared_ptr<DB::IAST>&) build_d
ocker/./src/Interpreters/InDepthNodeVisitor.h:51:13
    #19 0x562941dbca04 in DB::InDepthNodeVisitor<DB::(anonymous namespace)::ExplainAnalyzedSyntaxMatcher, true, false, std::__1::shared_ptr<DB::IAST>>::visit(std::__1::shared_ptr<DB::IAST>&) build_docker/./src/Inte
rpreters/InDepthNodeVisitor.h:32:13
    #20 0x562941dbca04 in DB::InterpreterExplainQuery::executeImpl() build_docker/./src/Interpreters/InterpreterExplainQuery.cpp:386:48
    #21 0x562941db9ae3 in DB::InterpreterExplainQuery::execute() build_docker/./src/Interpreters/InterpreterExplainQuery.cpp:91:20
    #22 0x562942c7d47d in DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/./src/Interpreters/executeQu
ery.cpp:1147:40
    #23 0x562942c75d26 in DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::E
num) build_docker/./src/Interpreters/executeQuery.cpp:1326:26
    #24 0x5629447f4c25 in DB::TCPHandler::runImpl() build_docker/./src/Server/TCPHandler.cpp:518:54
    #25 0x5629448271b5 in DB::TCPHandler::run() build_docker/./src/Server/TCPHandler.cpp:2312:9
    #26 0x562947ab8dbd in Poco::Net::TCPServerConnection::start() build_docker/./base/poco/Net/src/TCPServerConnection.cpp:43:3
    #27 0x562947ab9ef9 in Poco::Net::TCPServerDispatcher::run() build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:115:20
    #28 0x562947d25e4f in Poco::PooledThread::run() build_docker/./base/poco/Foundation/src/ThreadPool.cpp:188:14
    #29 0x562947d20f50 in Poco::ThreadImpl::runnableEntry(void*) build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:335:27
    #30 0x7f783cea4ac2 in start_thread nptl/pthread_create.c:442:8
```

I couldn't find the query that caused it, around this report queries from `02461_prewhere_row_level_policy_ligh
tweight_delete.gen.sql` where fuzzed.

Probably here `row_level_filter` is nullptr:
https://github.com/ClickHouse/ClickHouse/blob/fb02137fcc35799d9444d56c45db39afc68c131e/src/Interpreters/InterpreterSelectQuery.cpp#L2110C51-L2110C63

### PR DESCRIPTION
**Pull request by devcrafter**
Fixes https://github.com/ClickHouse/ClickHouse/issues/60762

Parallel replicas related code contributed to the issue in the case of `parallel_replicas_min_number_of_rows_per_replica` enabled. It enables additional query analysis if parallel replicas is enabled. If the query has a filter which can be moved to `PREWHERE` or has `PREWHERE`, `applyFiltersToPrewhereInAnalysis()` is executed, it modifies `ExpressionAnalysisResult::filter_info`. With non-zero `parallel_replicas_min_number_of_rows_per_replica`, `applyFiltersToPrewhereInAnalysis()` is called twice, and the second time it accesses `filter_info->actions` which triggers undefined behavior since the value was moved out of it during the first `applyFiltersToPrewhereInAnalysis()` call. The undefined behavior was introduced by https://github.com/ClickHouse/ClickHouse/pull/51692

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 1b2357198d4c1b92fbe3f5a6c0c11be6f1d80a8f with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61052/1b2357198d4c1b92fbe3f5a6c0c11be6f1d80a8f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -2122,7 +2121,7 @@ void InterpreterSelectQuery::applyFiltersToPrewhereInAnalysis(ExpressionAnalysis
     else
     {
         /// Add row level security actions to prewhere.
-        analysis.prewhere_info->row_level_filter = std::move(analysis.filter_info->actions);
+        analysis.prewhere_info->row_level_filter = analysis.filter_info->actions;
```
Comment (Algunenano) [inline src/Interpreters/InterpreterSelectQuery.cpp:2124]: What about next line? Should we also remove the std::move of `analysis.filter_info->column_name`

**Review by Algunenano — COMMENTED** [review]
It's taken me a while to understand why this fixes it as we check for `!analysis.filter_info`. The problem is here:

https://github.com/ClickHouse/ClickHouse/blob/2cea787f8163ae2ff522336614f463224d1bd2f5/src/Interpreters/InterpreterSelectQuery.cpp#L915-L916

I created a copy of `analysis_result` to be able to keep the code, but the copy is a "shallow" copy as most of the members of `ExpressionAnalysisResult` are shared pointers. So although it's a copy, the changes to its members are kept via the shared state.

Thanks for investigating and fixing it!

**Comment by devcrafter** [comment]
> It's taken me a while to understand why this fixes it as we check for `!analysis.filter_info`. The problem is here:
> 
> https://github.com/ClickHouse/ClickHouse/blob/2cea787f8163ae2ff522336614f463224d1bd2f5/src/Interpreters/InterpreterSelectQuery.cpp#L915-L916
> 
> I created a copy of `analysis_result` to be able to keep the code, but the copy is a "shallow" copy as most of the members of `ExpressionAnalysisResult` are shared pointers. So although it's a copy, the changes to its members are kept via the shared state.
> 
> Thanks for investigating and fixing it!

Yes, exactly. I noticed it during debugging, and the observation led me to the fix. Sorry, I didn't mention this detail in the description (thought it's already long enough)

**Comment by devcrafter** [comment]
Stress test failure - https://github.com/ClickHouse/ClickHouse/issues/61134

**Pull request merged**